keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,425 @@
+
+The metrics will be implemented via the collector pattern - when a scrape is requested the pod cache will be traversed and each metric will be calculated and reported in a streaming fashion, without caching or having to acquire locks. This is the standard pattern within Kubernetes for reporting high cardinality metrics and generally allows us to discount the cost of scrapes (estimate: one scrape every 15s might add 0.1 core of use for 50k pods, and we can control scape interval and reduce cost). The approach will add no additional memory usage overhead except for the transient allocations during iteration of the pod cache. Note that in other components we use this pattern to gather one or two orders of magnitude more metrics proportional to pods, so from a whole system perspective the impact is negligible and should have no impact on the scheduler performance.
+
+The series reported will be consistent with `kube_node_status_allocatable` to simplify queries that compare used capacity and available capacity. After a pod has reached the Completed lifecycle (as described above) state, the series will not be reported. If the value of the resource is zero, that resource will not report a series. They all share the form `kube_pod_resource_(requests|limits)`.
+
+Note on naming: The Prometheus convention for a metric is to allow the sum of a metric to make sense. However, in the case of Kubernetes resources we are attempting to make discoverability of the reported metrics (which are unbounded in number as they are extensible) a key goal for admins. Most admins will understand CPU and memory usage, but would have to resort to metrics name wildcard regexes to find the other values if we give each resource its own metrics name. Also, because Kube extended resources have characters disallowed in metrics names, admins would continually have to transform `myresource.io/foo-bar` into `myresource_io_foo_bar` in order to locate names which places an undue burden on the user. So for the proposed metrics and `kube_node_status_allocatable` we instead place all types resources under the same metrics name with a unit and resource label.  For this metric, the sum of series is only meaningful when filtering on a resource label.
+
+* `kube_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time.
+  * The value is the quantity of the request in `spec.containers.resources.requests`.
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled.
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `unit` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource or if it is an extension resource that does not have a detectable unit.
```

Comment:
adding pod priority as a label would be useful as well.

Reply:
I can add that, it seems reasonable and in keeping with the idea that priority implies resources.  Are there any other resource related metrics that have been requested for scheduling?

Reply:
Since it's going to be inside the scheduler, It would make sense to include the scheduler profile (which corresponds to Pod.spec.schedulerName) https://kubernetes.io/docs/reference/scheduling/config/#multiple-profiles

Reply:
But I'm not 100% convinced. I guess the same question applies to priority: how does it affect resource allocation? We don't have different resource calculations for different priorities or profiles.

Reply:
This can be useful to answer the question: how much resources are allocated for each priority level.

Reply:
+1 to adding the profile. Similarly, how much resources are allocated for each profile.

Reply:
Right.

+1 then :)

Can we say something similar for profiles? "How much a node is packed per profile?"

Reply:
Adding profile/schedulerName raised another question: should the in-tree (default) scheduler responsible for reporting the metrics of pod that is scheduled by out-of-tree schedulers?

Reply:
I believe yes because otherwise we can't answer the question: how much resources has been allocated.

Reply:
I say yes for scheduled Pods, as we take them into account for calculating remaining space. For unscheduled Pods, semantically we shouldn't.

Reply:
I would suggest that while the scheduler is hosting this, the metric is system wide.  There is no cost to reporting the metric for all profiles because while the metric may be interpreted differently, an admin still sees data that the scheduler sees the same way (the default schedulers’ decision not to act on those profiles doesn’t mean the api definition of those fields change)

Reply:
Fine with me. Worth adding the caveat in final Docs, that clusters administrators should ensure that only one scheduler is reporting metrics (if they have more than one leader).

Reply:
I added:

> Instrumentation for Kubernetes should scrape this endpoint for the default scheduler to satisfy the objectives around capacity management - more complex scheduling configurations are free to provide additional metrics that complement this data but are not required to duplicate the reporting or retrieval of this info.

Let me know if that satisfies.

Reply:
`more complex scheduling configurations, such as multiple schedulers running`

Reply:
Used "non-default schedulers are..." to avoid having to say schedulers twice

Reply:
sg